quickwit/quickwit-search/src/find_trace_ids_collector.rs

```diff
@@ -0,0 +1,508 @@
+            .map(BinaryHeap::from)
+            .map(binary_heap::IntoIterSorted::new)
+            .kmerge()
+        {
+            if !seen_trace_ids.contains(&trace_id.trace_id) {
+                seen_trace_ids.insert(trace_id.trace_id.clone());
+                trace_ids.push(trace_id);
+
+                if trace_ids.len() == self.num_traces {
+                    break;
+                }
+            }
+        }
+        Ok(trace_ids)
+    }
+
+    fn requires_scoring(&self) -> bool {
+        false
+    }
+}
+
+pub struct FindTraceIdsSegmentCollector {
+    trace_id_ff_reader: MultiValuedFastFieldReader<u64>,
+    span_timestamp_column: Arc<dyn Column<DateTime>>,
+    inverted_index_reader: Arc<InvertedIndexReader>,
+    select_trace_ids: SelectTraceIds,
+}
+
+impl FindTraceIdsSegmentCollector {
+    fn trace_id_term_ord(&self, doc: DocId) -> TermOrd {
+        self.trace_id_ff_reader
+            .get_first_val(doc)
+            .expect("There should be exactly one trace ID per span.")
+    }
+
+    fn span_timestamp(&self, doc: DocId) -> DateTime {
+        self.span_timestamp_column.get_val(doc)
+    }
+}
+
+impl SegmentCollector for FindTraceIdsSegmentCollector {
+    type Fruit = Vec<TraceIdFruit>;
+
+    fn collect(&mut self, doc: DocId, _score: Score) {
+        let term_ord = self.trace_id_term_ord(doc);
+        let span_timestamp = self.span_timestamp(doc);
+        self.select_trace_ids.collect(term_ord, span_timestamp);
+    }
+
+    fn harvest(self) -> Self::Fruit {
+        let mut trace_id_bytes = Vec::with_capacity(24);
+
+        self.select_trace_ids
+            .harvest()
+            .into_iter()
+            .map(|trace_id| {
+                let found = self.inverted_index_reader
+                    .terms()
+                    .ord_to_term(trace_id.term_ord, &mut trace_id_bytes)
+                    .expect("The term dict should be available.");
+                debug_assert!(found);
+                TraceIdFruit::new(&trace_id_bytes, trace_id.span_timestamp)
+            })
+            .collect()
+    }
+}
+
+struct SelectTraceIds {
+    num_traces: usize,
+    dedup_workbench: FnvHashMap<TermOrd, DateTime>,
+    select_workbench: Vec<TraceIdTermOrd>,
+    running_term_ord: Option<TermOrd>,
+    running_span_timestamp: DateTime,
+    span_timestamp_sentinel: DateTime,
+}
+
+impl SelectTraceIds {
+    fn new(num_traces: usize) -> Self {
+        Self {
+            num_traces,
+            dedup_workbench: FnvHashMap::with_capacity_and_hasher(
+                2 * num_traces,
+                Default::default(),
+            ),
+            select_workbench: Vec::with_capacity(2 * num_traces),
+            running_term_ord: None,
+            running_span_timestamp: DateTime::default(),
+            span_timestamp_sentinel: DateTime::from_timestamp_micros(i64::MIN),
+        }
+    }
+
+    fn collect(&mut self, term_ord: TermOrd, span_timestamp: DateTime) {
+        if self.running_term_ord.is_none() {
```

Comment:
nitpick...

I suspect the code could be midly cleaner with an Option<RunningTerm>

```
struct RunningTerm {
   term_ord,
   max_timestamp,
} 
```

You could then use `running_term_opt.get_or_insert()`